packages/three-vrm-springbone/src/VRMSpringBoneManager.ts

```diff
@@ -80,135 +89,123 @@ export class VRMSpringBoneManager {
     this.deleteJoint(joint);
   }
 
-  public setInitState(): void {
+  /**
+   * Sorts the joints ensuring they are updated in the correct order taking dependencies into account.
+   */
+  private _sortJoints() {
+    const springBoneOrder: Array<VRMSpringBoneJoint> = [];
     const springBonesTried = new Set<VRMSpringBoneJoint>();
     const springBonesDone = new Set<VRMSpringBoneJoint>();
-    const objectUpdated = new Set<THREE.Object3D>();
-
+    const ancestors: Array<THREE.Object3D> = [];
     for (const springBone of this._joints) {
-      this._processSpringBone(springBone, springBonesTried, springBonesDone, objectUpdated, (springBone) =>
-        springBone.setInitState(),
-      );
+      this._insertJointSort(springBone, springBonesTried, springBonesDone, springBoneOrder, ancestors);
     }
-  }
+    this._sortedJoints = springBoneOrder;
 
-  public reset(): void {
-    const springBonesTried = new Set<VRMSpringBoneJoint>();
-    const springBonesDone = new Set<VRMSpringBoneJoint>();
-    const objectUpdated = new Set<THREE.Object3D>();
-
-    for (const springBone of this._joints) {
-      this._processSpringBone(springBone, springBonesTried, springBonesDone, objectUpdated, (springBone) =>
-        springBone.reset(),
-      );
-    }
-  }
-
-  public update(delta: number): void {
-    const springBonesTried = new Set<VRMSpringBoneJoint>();
-    const springBonesDone = new Set<VRMSpringBoneJoint>();
-    const objectUpdated = new Set<THREE.Object3D>();
-
-    for (const springBone of this._joints) {
-      // update the springbone
-      this._processSpringBone(springBone, springBonesTried, springBonesDone, objectUpdated, (springBone) =>
-        springBone.update(delta),
-      );
-
-      // update children world matrices
-      // it is required when the spring bone chain is sparse
-      traverseChildrenUntilConditionMet(springBone.bone, (object) => {
+    const lca = lowestCommonAncestor(ancestors);
+    this._ancestors = [];
```

Comment:
I'm sure this `this._ancestors` has to be an array since the 0th element should calculate ancestors' world matrices.

[SHOULD] However, the reason why the variable `this._ancestors` is needed is kinda unintuitive. It would be very appreciated if you give them a proper doc comment to help future us.
I'm sure this `this._ancestors` has to be an array since the 0th element should calculate ancestors' world matrices.

[SHOULD] However, the reason why the variable `this._ancestors` is needed is kinda unintuitive. It would be very appreciated if you give them a proper doc comment to help future us.
I believe it goes like "A list of ancestors of the entire SpringBone joints. Their world matrices have to be calculated before we update the entire SpringBone system. The first element is ~~"

Reply:
Correct, the `_ancestors` list is ordered such that their world matrices can be updated correctly. The first element is the lowest common ancestor for which we also need to make sure its ancestors have their world matrix updated.

The graph below is an example of a Scene graph and shows which nodes would be in the `_ancestors` list in that case:
```mermaid
graph TD
    S[Scene] --> AP[Object3D]
    S[Scene] --> B[Object3D]
    AP --> A[Object3D<br>ancestors#91;0#93;]
    A --> C[Object3D<br>ancestors#91;1#93;]
    A --> D[Object3D<br>ancestors#91;2#93;]
    C --> S1[Springbone 1]
    D --> S2[Springbone 2]
    D --> S3[Springbone 3]
  classDef ancestor fill:#AFA,stroke:gray
  classDef lancestor fill:#4D4,stroke:black
  classDef o3d fill:#EEE,stroke:gray
  class C ancestor
  class D ancestor
  class A lancestor
  class S o3d
  class AP o3d
  class B o3d
```

I've added a doc comment that should hopefully clarify things for future readers.

Reply:
The doc comment looks perfect, thank you!